gnovm/pkg/gnolang/preprocess.go

```diff
@@ -2163,12 +2163,38 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				numNames := len(n.NameExprs)
 				sts := make([]Type, numNames) // static types
 				tvs := make([]TypedValue, numNames)
+
 				if numNames > 1 && len(n.Values) == 1 {
-					// special case if `var a, b, c T? = f()` form.
-					cx := n.Values[0].(*CallExpr)
-					tt := evalStaticTypeOfRaw(store, last, cx).(*tupleType)
-					if rLen := len(tt.Elts); rLen != numNames {
-						panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %s returns %d value(s)", numNames, cx.Func.String(), rLen))
+					// Special cases if one of the following:
+					// - `var a, b, c T = f()`
+					// - `var a, b = n.(T)`
+					// - `var a, b = n[i], where n is a map`
+
+					var tuple *tupleType
+					valueExpr := n.Values[0]
+					valueType := evalStaticTypeOfRaw(store, last, valueExpr)
+
+					switch expr := valueExpr.(type) {
+					case *CallExpr:
+						tuple = valueType.(*tupleType)
+					case *TypeAssertExpr, *IndexExpr:
+						tuple = &tupleType{Elts: []Type{valueType, BoolType}}
+						if ex, ok := expr.(*TypeAssertExpr); ok {
+							ex.HasOK = true
+							break
+						}
+						expr.(*IndexExpr).HasOK = true
+					default:
+						panic(fmt.Sprintf("unexpected ValueDecl value expression type %T", expr))
+					}
+
+					if rLen := len(tuple.Elts); rLen != numNames {
+						rhsExprString := valueExpr.String()
+						if cx, ok := valueExpr.(*CallExpr); ok {
+							rhsExprString = cx.Func.String()
+						}
```

Comment:
do we need this? seems `valueExpr.String()` is already good.

Reply:
This is just to keep it consistent with how this was working previously for call expressions -- CallExpr.String() does not return the same thing as CallExpr.Func.String().

Reply:
The difference between these two outputs is very small. If there aren't many side effects, I believe using one `valueExpr.String()` is better.